contracts/apps/accounting/AccountingApp.sol

```diff
@@ -0,0 +1,186 @@
+            value: value, 
+            // TODO: get base token and exchange rate from oracle 
+            baseToken: 0x0, 
+            baseValue: 1,
+            reference: reference, 
+            timestamp: now,
+            accountingPeriodId: getCurrentAccountingPeriodId()
+        })) - 1;
+        // All transactions must have at least one state.
+        // To optimize, incoming transactions could go directly to "Suceeded" or "Failed".
+        updateTransaction(tid, initialState, "new");
+        return tid;
+    }
+
+
+    // Create new transactionUpdate for the given transaction id
+    bytes4 constant UPDATE_TRANSACTION_SIG = bytes4(sha3('updateTransaction(uint,TransactionState,string)'));
+    function updateTransaction(uint transactionId, TransactionState state, string reason) returns (uint) {
+        uint tuid = transactionUpdates.push(TransactionUpdate({
+            transactionId: transactionId,
+            state: state,
+            reason: reason,
+            actor: msg.sender
+        })) - 1;
+        transactionUpdatesRelation[transactionId].push(tuid);
+        return tuid;
+    }
+
+    bytes4 constant SET_TRANSACTION_SUCCEEDED_SIG = bytes4(sha3('setTransactionSucceeded(uint,string)'));
+    function setTransactionSucceeded(uint transactionId, string reason) {
+        updateTransaction(transactionId, TransactionState.Succeeded, reason);
+    }    
+
+    bytes4 constant SET_TRANSACTION_PENDING_APPROVAL_SIG = bytes4(sha3('setTransactionPendingApproval(uint,string)'));
+    function setTransactionPendingApproval(uint transactionId, string reason) {
+        updateTransaction(transactionId, TransactionState.PendingApproval, reason);
+    }
+
+    bytes4 constant SET_TRANSACTION_FAILED_SIG = bytes4(sha3('setTransactionFailed(uint,string)'));
+    function setTransactionFailed(uint transactionId, string reason) {
+        updateTransaction(transactionId, TransactionState.Failed, reason);
+    }
+
+    // This flattens the last TransactionUpdate with the base Transation to show the current state of the transaction.
+    // This assumes that there is at least a single transaction update which is fine if newTransaction is used.
+    bytes4 constant GET_TRANSACTION_STATE_SIG = bytes4(sha3('getTransactionState(uint)'));
+    function getTransactionState(uint transactionId) constant returns (address token, int value, string reference, uint timestamp, string stringState, TransactionState state, uint accountingPeriodId) {
+        Transaction t = transactions[transactionId];
+        uint lastTransactionUpdate = transactionUpdatesRelation[transactionId][transactionUpdatesRelation[transactionId].length - 1];
+        TransactionUpdate tu = transactionUpdates[lastTransactionUpdate];
+        token = t.token;
+        value = t.value;
+        reference = t.reference;
+        timestamp = t.timestamp;
+        state = tu.state;
+        accountingPeriodId = t.accountingPeriodId;
+        if (state == TransactionState.Succeeded) {
```

Comment:
I wouldn't return strings from the contract. Should really be a frontend concern.